### PR TITLE
Change C++ version to C++17

### PR DESCRIPTION
**Pull request by esteffin**
Changed gnu-make and cmake-based builds to C++17.

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by tautschnig** [comment]
What about #7172?

File: src/goto-instrument/document_properties.cpp [hunk 1]
```diff
@@ -238,7 +238,7 @@ document_propertiest::get_code(const source_locationt &source_location)
       while(line_no.size()<4)
         line_no=" "+line_no;
 
-      line_no+"  ";
+      line_no += "  ";
```
Comment (tautschnig) [inline src/goto-instrument/document_properties.cpp]: Seems unrelated?!

Comment (thomasspriggs) [reply src/goto-instrument/document_properties.cpp]: I agree that this is worth separating into its own PR. Some of the new versions of the standard library add the `[[nodiscard]]` attribute to the + operator of `std::string`. This causes a compile error when the result of the concatenation is unused. It essentially promotes a logic error into a compile error. Without this commit, the `check-ubuntu-22_04-cmake-gcc-13` job will fail, so it does appear to be needed before we can switch standards. However as it is a logic error currently we could fix it before switching standards.

Comment (esteffin) [reply src/goto-instrument/document_properties.cpp]: That is a fix for an issue brought up by the C++17 change as the `operator+` between strings is marked `nodiscard` and so it was resulting in a compilation error.

If it is too awkward to include this here, I'm happy to move the change to a separate PR, but I would need it in before upgrading the standard.

Comment (tautschnig) [reply src/goto-instrument/document_properties.cpp]: I think this should be PR'ed and fixed right away. (I guess no one is using that feature so it went unnoticed, but it should really be fixed now.)

Comment (thomasspriggs) [reply src/goto-instrument/document_properties.cpp]: @tautschnig I just read the surrounding code and the value of `line_no` is never read back and output. So it is just as broken before and after this commit...

**Comment by martin-cs** [comment]
We have traditionally been very conservative about C++ version increments as it is not uncommon to find the "approved" Linux platform some users have trailing the current version by as much as 5 years.  Ubuntu 18.04 LTS is only just out of general support, still gets security fixes and I strongly suspect still has quite a user base.  Generally supporting an older C++ version is a good thing.  But, it can be worth increasing the version if it makes the project better in some way.  From this PR it is not clear to me what the benefits are.  There are probably some that I just don't see.  Can anyone say what they are?

**Comment by thomasspriggs** [comment]
> We have traditionally been very conservative about C++ version increments as it is not uncommon to find the "approved" Linux platform some users have trailing the current version by as much as 5 years. Ubuntu 18.04 LTS is only just out of general support, still gets security fixes and I strongly suspect still has quite a user base. Generally supporting an older C++ version is a good thing. But, it can be worth increasing the version if it makes the project better in some way. From this PR it is not clear to me what the benefits are. There are probably some that I just don't see. Can anyone say what they are?

@martin-cs Just a reminder - you approved the previous PR making the equivalent changes and you seemed satisfied with the motivating factors at the time - https://github.com/diffblue/cbmc/pull/7172#pullrequestreview-1128557329

A quick summary of some of the advantages off the top of my head -

- Removing the library code which has been back ported C++11, which will reduce the maintenance needed.
- Using standard library versions of code lowers the barriers to entry for experienced C++ developers who are not yet familiar with the cbmc. Currently contributors who fall into this category have to learn details such as use `optionalt` instead of `std::optional`.
- Access to new C++17 STL classes such as `std::variant`.

File: src/goto-instrument/document_properties.cpp (hunk 1 above)
Comment (tautschnig) [reply src/goto-instrument/document_properties.cpp]: Fair point. I suspect this should have been inserted as a prefix into each line of output. Let me make a stab at a PR.

**Comment by martin-cs** [comment]
@thomasspriggs thanks for the reminder; had totally forgotten about that PR!

**Review by martin-cs — APPROVED** [review]
Approve subject to @tautschnig 's comments about separate PRs.

File: src/goto-instrument/document_properties.cpp (hunk 1 above)
Comment (tautschnig) [reply src/goto-instrument/document_properties.cpp]: Fix in #7994.

Comment (esteffin) [reply src/goto-instrument/document_properties.cpp]: I will remove this commit from this PR as soon as the latter is merged.

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,5 +1,9 @@
 cmake_minimum_required(VERSION 3.8)
 
+set(CMAKE_CXX_STANDARD 17)
+set(CMAKE_CXX_STANDARD_REQUIRED true)
+set(CMAKE_CXX_EXTENSIONS OFF)
```
Comment (NlightNFotis) [inline CMakeLists.txt:5]: Thank the stars we're not depending on any extensions from before (when I first saw this change I nearly had a heart attack, but then I noticed CI passing and I let out a sigh of relief).

**Review by NlightNFotis — APPROVED** [review]

**Review by tautschnig — APPROVED** [review]
